### PR TITLE
Add support for Wayland security context

### PR DESCRIPTION
**Pull request by emersion**
This exposes a reliable way for Wayland compositors to get
identifying information about a client. Compositors can then
apply security policies if desirable.

~~TODO: `bwrap --sync-fd` can only be specified once (only the last param is taken into account), but D-Bus already uses it. Any ideas how to address this?~~

References: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/68

### CODE REVIEW

File: common/flatpak-run.c
Comment (alexlarsson) [inline common/flatpak-run.c:484]: Is this really needed? I think wayland will find a default display anyway, no?

Comment (alexlarsson) [reply common/flatpak-run.c:484]: Yeah, the other codepath defaults to wayland-0 in this case

File: common/flatpak-run.c
Comment (alexlarsson) [inline common/flatpak-run.c:518]: I don't see why you can't reuse the existing sync-fd with the dbus proxy. Just needs some shared code in FlatpakBwrap to handle the allocation of it.

File: common/flatpak-run.c
Comment (emersion) [reply common/flatpak-run.c]: It's to avoid connecting to a random Wayland compositor when running outside of a Wayland session. But oh well, that ship has already sailed. Removed.

File: common/flatpak-run.c
Comment (emersion) [reply common/flatpak-run.c]: Done!

**Comment by emersion** [comment]
Gentle ping

File: common/flatpak-run.c
Comment (swick) [inline common/flatpak-run.c:558]: If the compositor supports the protocol and we still fail to create the context then shouldn't we error out here instead of giving the app unconstrained access?

I already don't know how the compositor would even know if flatpak is old or the app is from the host.

**Comment by swick** [comment]
@smcv might also be interested

File: common/flatpak-run.c
Comment (emersion) [reply common/flatpak-run.c]: Fixed.

File: common/flatpak-run.c
Comment (jadahl) [inline common/flatpak-run.c:495]: I wonder if any of these needs to be added for `apply_extra_data()` (which is where `instance_id` can be `NULL`).

**Comment by emersion** [comment]
Rebased, added Meson support, made instance ID mandatory.

File: common/flatpak-run.c
Comment (swick) [reply common/flatpak-run.c:495]: Mh, I would even go so far and say it also should not get access to IPC, /dev, dri and shared tmp.

Probably makes sense to separate into a new PR.

**Comment by davidedmundson** [comment]
Can I check that this MR is "acked in principle" and just blocked on  the upstream protocol to land?

Obviously it's chicken and egg, but it's also silly to land this in the specs without the most important user of this being officially on-board.

**Comment by jadahl** [comment]
> Can I check that this MR is "acked in principle" and just blocked on the upstream protocol to land?

There has been no explicit ack from a maintainer yet, but it got some initial review from @alexlarsson. Something more would be appreciated no doubt.

**Comment by emersion** [comment]
Any news about this?

**Review by alexlarsson — APPROVED** [review]
LGTM

**Comment by emersion** [comment]
Thank you for having a look, @alexlarsson!

**Comment by jadahl** [comment]
Next steps: release wayland-protocols with the new protocol, and bump the version needed by flatpak if enabled.

File: common/flatpak-run.c
Comment (alexlarsson) [reply common/flatpak-run.c:495]: Generally these are no-ops due to the passed in run_flags (e.g. with FLATPAK_RUN_FLAG_NO_SESSION_BUS_PROXY, etc), and the empty app_context that apply_extra_data uses.

Do you have any particular place where it gets this wrong?

Comment (alexlarsson) [reply common/flatpak-run.c:495]: In fact, is this check even needed? I don't see how `sockets & FLATPAK_CONTEXT_SOCKET_WAYLAND` would be set in apply_extra_data?

In fact, isn't it a bad idea to not call flatpak_run_add_socket_args_environment()?
We actively do some things like cover-mount /tmp/.X11, etc,  if the x11 flag is not set. So maybe this will actually limit our security measures?

Comment (emersion) [reply common/flatpak-run.c:495]: I see. So instead, would an assert that the instance is not NULL be welcome when setting up Wayland sockets?

Comment (jadahl) [reply common/flatpak-run.c:495]: In the end, what we don't is setting up Wayland socket when we don't have an instance id, which appears to be the case when we arrive here from `apply_extra_data()`.

Comment (jadahl) [reply common/flatpak-run.c:495]: > I see. So instead, would an assert that the instance is not NULL be welcome when setting up Wayland sockets?

I guess this makes sense. Assert that instance ID is set if `sockets & FLATPAK_CONTEXT_SOCKET_WAYLAND`.

**Comment by jadahl** [comment]
> Next steps: release wayland-protocols with the new protocol, and bump the version needed by flatpak if enabled.

Can now depend on 1.32 for this.

File: common/flatpak-run.c
Comment (swick) [reply common/flatpak-run.c:495]: I looked into this as well on Friday and noticed that the run flags are not what makes this work. The context doesn't load metadata from anywhere so none of the sockets will get set up.

The `--sandboxed` argument to `flatpak run` was added after the `apply_extra_data` sandbox and does almost the same but not entirely. It does add `FLATPAK_RUN_FLAG_NO_PROC` as well and I can't tell why and we don't call `flatpak_context_make_sandboxed` which is fine because we don't load anything else in the context but might be a good idea to call anyway.

Wanted to test this but a lot of tests are failing for me on main because of #5027.

```patch
--- a/common/flatpak-dir.c
+++ b/common/flatpak-dir.c
@@ -8328,20 +8328,25 @@ apply_extra_data (FlatpakDir   *self,
                           "--cap-drop", "ALL",
                           NULL);
 
-  /* Might need multiarch in apply_extra (see e.g. #3742).
-   * Should be pretty safe in this limited context */
-  run_flags = (FLATPAK_RUN_FLAG_MULTIARCH |
+  /* Run flags which equal --sandboxed */
+  run_flags = (FLATPAK_RUN_FLAG_SANDBOX |
                FLATPAK_RUN_FLAG_NO_SESSION_HELPER |
-               FLATPAK_RUN_FLAG_NO_PROC |
                FLATPAK_RUN_FLAG_NO_SESSION_BUS_PROXY |
                FLATPAK_RUN_FLAG_NO_SYSTEM_BUS_PROXY |
                FLATPAK_RUN_FLAG_NO_A11Y_BUS_PROXY);
 
+  /* Might need multiarch in apply_extra (see e.g. #3742).
+   * Should be pretty safe in this limited context.
+   * Removing /proc to tighten sandbox even more I guess? */
+  run_flags |= FLATPAK_RUN_FLAG_NO_PROC |
+               FLATPAK_RUN_FLAG_MULTIARCH;
+
   if (!flatpak_run_setup_base_argv (bwrap, runtime_files, NULL, runtime_arch,
                                     run_flags, error))
     return FALSE;
 
   app_context = flatpak_context_new ();
+  flatpak_context_make_sandboxed (app_context);
 
   if (!flatpak_run_add_environment_args (bwrap, NULL, run_flags, id,
                                          app_context, NULL, NULL, -1,
```

**Comment by emersion** [comment]
Added an assert and the wayland-protocols version requirement.

File: common/flatpak-run.c
Comment (alexlarsson) [reply common/flatpak-run.c:494]: We absolutely don't want NO_PROC. That will break any apps that uses /proc/self/fd, etc which is very common (even used by glibc).  And /proc access is not a security problem for regular apps. Only for apply_extra_data() when writing to the system repo. See commit message for cd2142888fc4c199723a0dfca1f15ea8788a5483.

And flatpak_context_make_sandboxed() is a complete no-op for the newly created (and thus empty) app_context.

File: common/flatpak-run-wayland.c
Comment (swick) [inline common/flatpak-run-wayland.c:121]: This should either be moved over the `socket` call (but I think this is as close to the `bind` call as possible) or must be handled at `out`.

File: .github/workflows/check.yml
Comment (swick) [inline .github/workflows/check.yml:205]: Why is this disabled in CI? Why is this a build time in the first place?

Comment (alexlarsson) [reply .github/workflows/check.yml:205]: Maybe the default should be "auto".

Comment (emersion) [reply .github/workflows/check.yml]: I changed the default to "auto".

File: common/flatpak-run-wayland.c
Comment (emersion) [reply common/flatpak-run-wayland.c]: Moved it over `socket()`.

**Comment by swick** [comment]
LGTM

File: common/flatpak-run-wayland.c
Comment (swick) [inline common/flatpak-run-wayland.c:119]: So, we're leaking listen_fd in the success case. This is generally not too bad, but if we want to close it, where can we do so safely? I guess after `wp_security_context_manager_v1_create_listener` and `wl_display_roundtrip` have been issued the fd is on the other end and will stay open.

Comment (swick) [reply common/flatpak-run-wayland.c:119]: we're not. never mind.

Comment (emersion) [reply common/flatpak-run-wayland.c:119]: This has been discussed in https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/232#note_1991868

**Comment by emersion** [comment]
@alexlarsson, anything else I need to do or does this look good to you?

**Comment by emersion** [comment]
@alexlarsson, any news about this?

File: common/flatpak-run-wayland.c
Comment (zccrs) [inline common/flatpak-run-wayland.c:187]: Why security context create failed will return FALSE?
Though `security_context_available` is true, if still using wayland mode, what will happen?.

Comment (emersion) [reply common/flatpak-run-wayland.c:187]: The goal here is to bail out if the security context protocol is supported by the compositor but we didn't manage to set it up. If the security context protocol is not supported, we want to carry on.

Comment (emersion) [reply common/flatpak-run-wayland.c:187]: No, continuing would be insecure, potentially running the client without a security context.

I don't think Flatpak can run inside Flatpak, so this doesn't really matter.

**Comment by emersion** [comment]
@alexlarsson, any chance to get this merged?

File: common/flatpak-run-wayland.c
Comment (alexlarsson) [inline common/flatpak-run-wayland.c:44]: Indentation is wrong

**Review by alexlarsson — APPROVED** [review]
Some very minor indentation issues, but otherwise this looks good to me,

File: common/flatpak-run-wayland.c
Comment (orowith2os) [inline common/flatpak-run-wayland.c:188]: I'd appreciate a bit of an explanation on what's going on here; the `if` check seems a bit backwards. AIUI, this would mean not running with Wayland at all, thus running *on X11*; why is that desired? You'd be breaking things like fractional scaling and mixed monitor scales when doing so. And even a Wayland socket *without* a security_context is leaps better than an X11 socket on its own.

Comment (orowith2os) [reply common/flatpak-run-wayland.c:190]: I'm still not sure on *why* we'd not prefer to use Wayland. What if a compositor doesn't support the protocol (like an old Mutter) but doesn't provide any iffy protocol extensions? Then this would mean all Flatpak apps run on X11 (if Flatpak is built with security_context support), which would result in a degraded UX. 

It seems more appropriate to log the issue and continue on.

Comment (emersion) [reply common/flatpak-run-wayland.c]: If the compositor doesn't support the protocol, the current logic will just set up a regular Wayland connection. No X11 used.

Comment (orowith2os) [reply common/flatpak-run-wayland.c]: It seems the general flow of the code was confusing me. There's `flatpak_run_add_wayland_security_context_args`, which does all of the security_context *and* wayland socket setup, and `flatpak_run_add_wayland_args`, which does the normal wayland stuff.

`flatpak_run_add_wayland_args` calls `flatpak_run_add_wayland_security_context_args` to completely set up the socket from here. If the protocol is available but failed to be set up, it'll bail out, running on X11. I would still prefer that to instead log an error and run on Wayland, but w/e. 

But I'm not understanding how the rest of the `flatpak_run_add_wayland_args` function is called if the compositor doesn't support the protocol. Is the rest of the function that falls back to that behavior done in `flatpak_run_add_wayland_security_context_args`? If so, would be nice to document that.

The confusion comes from the variable being named `security_context_available` - the final `if` check on it makes it *sound* like it's "if security context is available, fail to set up Wayland", when in reality it's just returning the initial value of the variable, `FALSE`. 

It would be nice if the variable could be set to `TRUE` initially instead, the security context setup has an `else` case that sets it to `FALSE`, and *then* the final check goes to see if the variable is *not* true, and returns `FALSE` in that case. That would make it flow a bit nicer. 

Also documenting the functions, their flow, and their behavior would be nice, to make it so conversations like this don't crop up again.

Comment (emersion) [reply common/flatpak-run-wayland.c]: Sorry, I don't have plans to address your comments. I think the current logic is fine, especially with the comment.

Comment (orowith2os) [reply common/flatpak-run-wayland.c]: The comment does help, and at most, I'd make it a tiny bit more explicit in the event the compositor doesn't support it:

```c
 /* If security-context is available but we failed to set it up, bail out.
  * If the compositor doesn't support security-context, we'll just
  * set up a plain Wayland socket.
  */
 ```

Comment (jadahl) [reply common/flatpak-run-wayland.c]: That's just repeating what the code already say.

It's well known convention in flatpak that returning true means "success", so if something say

```
if (try_security_context(&was_it_even_there))
  return true;
if (was_it_even_there)
  return false;
```

is very clear. The only question is "why", not "what", which could just as well be in the commit message.

**Comment by emersion** [comment]
Thanks a lot!

**Pull request merged**
